include/package-pack: actually call default_postinist/prerm with APK

robimarko opened the pull request:
Currently, when you install a package that ships uci-defaults scripts they are not getting executed during package installation.

For example, if you install a new LuCI theme it ships a uci-defaults script to add it to the UCI LuCI config so the theme is actually selectable but that does not happen until the device is rebooted so that uci-defaults script is actually executed.

It turns out that the recipe for post-install is the issue since it will include contents of postinst-pkg before the call to default_postinist. This is an issue since postinst-pkg will exit with code 0 before we ever reach the call to default_postinist.

So, lets simply make the call to default_postinist before postinst-pkg inclusion which is what OPKG based installation does.

The same issue affects pre-deinstall script but its not actually being hit since we dont seem to be actually even generating prerm-pkg scripts but lets fix this theoretical issue as well.

Fixes: #16987
Fixes: d788ab376f85 ("build: add APK package build capabilities")

robimarko committed:
include/package-pack: actually call default_postinist/prerm when using APK

Currently, when you install a package that ships uci-defaults scripts they
are not getting executed during package installation.

For example, if you install a new LuCI theme it ships a uci-defaults script
to add it to the UCI LuCI config so the theme is actually selectable but
that does not happen until the device is rebooted so that uci-defaults
script is actually executed.

It turns out that the recipe for post-install is the issue since it will
include contents of postinst-pkg before the call to default_postinist.
This is an issue since postinst-pkg will exit with code 0 before we ever
reach the call to default_postinist.

So, lets simply make the call to default_postinist before postinst-pkg
inclusion which is what OPKG based installation does.

The same issue affects pre-deinstall script but its not actually being hit
since we dont seem to be actually even generating prerm-pkg scripts but
lets fix this theoretical issue as well.

Fixes: #16987
Fixes: d788ab376f85 ("build: add APK package build capabilities")
Link: https://github.com/openwrt/openwrt/pull/17023
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@ynezz @aparcar For OPKG we rely on `default_prerm` and `default_postinst` to call the relevant `prerm-pkg` and `postinst-pkg` scripts instead of calling them directly in `pre-deinstall` and `post-install` but I guess that this is a APK limitation.

vernesong commented:
postrm seems also not work

robimarko commented:
@vernesong Why do you think so?
I just tested removing an LuCI theme and it runs postrm just fine since its removing the UCI entries like stated in its postrm script

vernesong commented:
> @vernesong Why do you think so? I just tested removing an LuCI theme and it runs postrm just fine since its removing the UCI entries like stated in its postrm script

I installed a build here : https://downloads.immortalwrt.org/snapshots/targets/x86/64/
and I test a luci package by compile with the codes below 
```
define Package/$(PKG_NAME)/postrm
#!/bin/sh
  echo "123" >>/tmp/1
  exit 0
endef

After I use the apk del command there is no file named 1 find in /tmp

robimarko commented:
No idea, but `postrm` works for me, it clearly gets executed upon removal of for example `luci-theme-material`, it did not work before this PR though so you cannot just prebuilts.

robimarko commented:
Hm, now I am wondering how it works for `luci-theme-material` but for my for test virtual package it doesn't get included at all, there is literally no postrm script.

robimarko commented:
Ok, so APK properly supports post install, post deinstall etc states, but make is driving me nuts as I cannot figure out how to make it add the appropriate --script argument to APK only if the file actually exists, it either doesn't add the argument or always evaluates true

robimarko committed:
include/package-pack: add support for postrm with APK

Currently, postrm scripts are not being executed when package is removed
with APK since they are not being included at all.

So, lets use APK-s support for post-deinstall scripts and ship our postrm
scripts if packages define them.

Fixes: d788ab376f85 ("build: add APK package build capabilities")
Link: https://github.com/openwrt/openwrt/pull/17023
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
`postrm` should also work now with the last push

aparcar commented:
Awesome thanks for taking care!

efahl commented:
I just tested by adding postinst and postrm to an existing package, rebuilding all and examining the contents of the apk.  All looked correct to me with the script contents in the right places.  `scp`ed the apk to my test VM, did a couple successive `add` and `del` sequences, output was generated correctly from both:

```
define Package/$(PKG_NAME)/postrm
#!/bin/sh
  echo "postrm"
  exit 0
endef
define Package/$(PKG_NAME)/postinst
	echo "postinst"
endef
```

robimarko commented:
Thanks! Rebased on top of main and merged!
